protocols/raft/src/main/java/io/atomix/protocols/raft/storage/snapshot/FileSnapshot.java

```diff
@@ -65,32 +69,57 @@ public synchronized SnapshotReader openReader() {
     return openReader(new SnapshotReader(buffer.mark().limit(SnapshotDescriptor.BYTES + Integer.BYTES + length), this), descriptor);
   }
 
-  @Override
-  public boolean isPersisted() {
-    return true;
-  }
-
   @Override
   public Snapshot complete() {
-    Buffer buffer = FileBuffer.allocate(file.file(), SnapshotDescriptor.BYTES);
+    checkNotNull(file.temporaryFile(), "no temporary snapshot file to read from");
+
+    Buffer buffer = FileBuffer.allocate(file.temporaryFile(), SnapshotDescriptor.BYTES);
     try (SnapshotDescriptor descriptor = new SnapshotDescriptor(buffer)) {
       descriptor.lock();
     }
+
+    try {
+      Files.move(file.temporaryFile().toPath(), file.file().toPath());
```

Comment:
Can we first try with StandardCopyOption.ATOMIC_MOVE then fallback with no option?

Reply:
Assuming it works with TMP files.

Reply:
I think the main issue here is that `ATOMIC_MOVE` doesn't work well across file systems where it needs to actually be copied; I think setting up your system temp directory as something like `ramfs` or `tmpfs` is fairly common, so I would expect it to fail. We can keep the current solution as fallback, but I guess I would expect the fallback to be used more often; I have no proof for this though.

Reply:
Yeah that's where I was going with my previous comment but I'm traveling right now so difficult to verify.

Reply:
I verified on my own system where `/tmp` is mounted as a `tmpfs` partition and the Raft log is on a different standard `ext4` partition. What I meant by no proof is I don't really know that most production setups will have `tmp` on a different partition with a different file system, but if I had to guess I'd say most likely.

Reply:
I suppose we could pass in the third argument directory and use the storage path along with StandardCopyOption.ATOMIC_MOVE and fallback to cut down on IO?

Reply:
It might be better to use the storage path anyway, as large snapshots could eat up quite a lot of RAM if your temp directory is mounted using `tmpfs`. I guess we can use UUIDs to keep the path unique and avoid two instances from two different threads writing to the same file (since afaik file locks are only helpful when dealing with multiple JVMs). Or do you have a another idea?

Reply:
sgtm, but I am not sure if the uuid is necessary? the first arg is prefix, Java adds some extra random chars with createTemp.. https://docs.oracle.com/javase/7/docs/api/java/io/File.html#createTempFile(java.lang.String,%20java.lang.String,%20java.io.File)

Reply:
Yeah you're right, seems pretty unlikely you'd get collisions.